### PR TITLE
Use node_ancestors to get rid of some while-loops

### PR DESCRIPTION
**Pull request by nickdrozd**
|     | Type                   |
| --- | ---------------------- |
| ✓   | :hammer: Refactoring   |

### CODE REVIEW

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
Can we keep only the one that are clear cut simplification where we replace 3 lines by a for loops ? Imo it's harder to understand a break in the for loop than the original whiles and the one where we have to still do ``node_origin = parent`` became really confusing.

**Comment by nickdrozd** [comment]
Refactoring these loops turned out to be a little trickier than I anticipated. The logic is broadly similar, but there are lots of small differences and irregularites. My feeling is that these differences are mostly accidental, but I don't know how to deal with them just now.

File: pylint/checkers/typecheck.py
Comment (nickdrozd) [inline pylint/checkers/typecheck.py:548]: @Pierre-Sassoulas Is this the one you're talking about? I agree that it's a little complicated, but I would say that was _already_ the case, especially with this double-assignment parent-child walk-up maneuver here. Getting rid of that IMO justifies adding in the break logic. But I can take it out if you feel strongly about it.

Maybe a way to make it clearer would be to return false straightaway instead of breaking the loop.

Comment (Pierre-Sassoulas) [reply pylint/checkers/typecheck.py:548]: I don't have strong opinion, let's see what Marc think.

File: pylint/__pkginfo__.py
Comment (cdce8p) [inline pylint/__pkginfo__.py:30]: IMO the `while` statement is a bit easier to read and understand.

File: pylint/checkers/base.py
Comment (cdce8p) [inline pylint/checkers/base.py:839]:
```suggestion
        for parent in node.node_ancestors():
```
I would suggest to be consistent with it and use `parent` whenever possible.

File: pylint/checkers/base.py
Comment (cdce8p) [inline pylint/checkers/base.py:1487]:
```suggestion
        for parent in node.node_ancestors():
```

File: pylint/checkers/typecheck.py
Comment (cdce8p) [reply pylint/checkers/typecheck.py:548]: It works either way, with `while` and with `for`. However I think the previous code was a bit easier to understand in this case. I especially like the tuple assignment as it makes it clear what actually happens.

If I remember correctly, I used something similar in another helper method.

--
A change to an early `return` wouldn't change much. Instead it would only create the possibility to add new errors.

File: pylint/checkers/utils.py
Comment (cdce8p) [inline pylint/checkers/utils.py:378]:
```suggestion
    for parent in var_node.node_ancestors():
```

**Review by cdce8p — CHANGES_REQUESTED** [review]
> Imo it's harder to understand a break in the for loop than the original whiles

It depends on the situation. For the changes here, I think it's mostly ok to use an early break. In particular something like this is quite easy to understand
```py
for parent in node.node_ancestors():
    if isinstance(parent, breaker_classes):
        break
    ...
```

> Refactoring these loops turned out to be a little trickier than I anticipated. The logic is broadly similar, but there are lots of small differences and irregularites. My feeling is that these differences are mostly accidental, but I don't know how to deal with them just now.

Noticed that as well while looking through the code. Unfortunately, I don't think it's accidental. Each and ever one of the checks / helper methods needs to deal with a very specific set of rules. Something simple like where the loop starts, at the current node or the parent, can have a huge impact and potentially break stuff.

--
The new helper is good, but we don't need to force it upon the existing code. Only using it where it makes sense is probably a good compromise.

**Review by Pierre-Sassoulas — APPROVED** [review]
I agree with @cdce8p's remarks, once they're taken into account we can merge.

File: pylint/__pkginfo__.py
Comment (nickdrozd) [reply pylint/__pkginfo__.py]: I tried another way with no loop at all. Compare this with https://github.com/PyCQA/pylint/pull/4860/files#diff-26d7b9f740481d02edafcd71a3f448d5d2f5a7ff9858bfa59622e72cfc87a3ed

File: pylint/checkers/base.py
Comment (nickdrozd) [reply pylint/checkers/base.py]: My first thought was to make the changes while disturbing the rest of the code as little as possible. But the variables are renamed now.

File: pylint/__pkginfo__.py
Comment (cdce8p) [inline pylint/__pkginfo__.py:31]: Just my opinion, but I would prefer the old way. I find it more intuitive.
Besides, I don't see a good point in changing a method that is only kept for compatibility reasons.

File: pylint/checkers/base.py
Comment (cdce8p) [inline pylint/checkers/base.py:1493]: Iterating through `node.parent` and `node.parent.parent` seems to be more common than I thought. Maybe we should add a utils method for it, this time in `pylint/checkers/utils.py`. Something like `walk_node_parents` or `recurse_node_parents`?

File: pylint/checkers/typecheck.py
Comment (cdce8p) [reply pylint/checkers/typecheck.py:548]: --
Obviously, this would be a good spot for the new helper.

File: pylint/checkers/variables.py
Comment (cdce8p) [inline pylint/checkers/variables.py:1352]: It could be used here as well

**Review by cdce8p — CHANGES_REQUESTED** [review]

File: pylint/__pkginfo__.py
Comment (nickdrozd) [reply pylint/__pkginfo__.py]: Lately I've been exploring the idea of programming using only primitive-recursive operators. Turing-completeness is not often required, and most tasks can be done without it. The `while` operator is awfully powerful, and it's definitely overkill for such a simple task as "pad out this list until it's at least of length 3". Ideally there would be no `while` loops at all except for where they're really needed, like perhaps in inference.

Of course this is an idiosyncratic concern, so I reverted the change.

File: pylint/checkers/base.py
Comment (nickdrozd) [reply pylint/checkers/base.py:1493]: There's another common variation where the last ancestor satisfying some condition is returned. I'm not sure if that needs another method or if there's a different pattern for that kind of thing.

File: pylint/__pkginfo__.py
Comment (Pierre-Sassoulas) [reply pylint/__pkginfo__.py]: There's this french author, Perec, who wrote a full novel without the letter e in it. It's an awfully powerful letter - the most used one in french in fact - and apparently, often you don't really need it 😄 Anyway I digress, removing part of a language for the beauty of it is interesting as an exercise, but pylint's users won't benefit from it and may even suffer from it because reviewing opinionated changes takes maintainer time that could be used elsewhere (we have 690 issues currently). Please don't refactor working code without a good reason.

File: pylint/checkers/base.py
Comment (Pierre-Sassoulas) [reply pylint/checkers/base.py:1493]: I think a generator ``n for n in node.node_ancestor if satisfy_condition(n)`` is the best solution in term of readability for those.

Comment (nickdrozd) [reply pylint/checkers/base.py:1493]: Yet another variation is looping over a node's siblings rather than its ancestors.

Comment (cdce8p) [reply pylint/checkers/base.py:1493]: If we think about this even longer, I'm sure we would be able to come up with further variations.

Personally, I would like to finish this PR soon. It has taken a bit too much time already.
To keep the spirit of @Pierre-Sassoulas last comment https://github.com/PyCQA/pylint/pull/5005#discussion_r710195942, I suggest we keep these three instance as they are now. It works, so why change it. I understand that probably isn't what you were hoping for @nickdrozd, but I believe this will be the easiest (and safest) solution.

--
I'm referring only to those that use child + parent recursively.

**Comment by nickdrozd** [comment]
Okay, I pared it down to just the straightforward cases. Thanks for reviewing!

**Review by cdce8p — APPROVED** [review]
LGTM!
Thanks @nickdrozd 🐬

**Pull request merged**
